yetanotherco/aligned_layer
# feat: operator liveness metric

# Operator Liveness Metric

## Motivation

We need a way to rapidly determine if an operator is down.

## Description

Adds a Bar Gauge with the count of missed tasks for each operator over a specified time range.

![image](https://github.com/user-attachments/assets/23f60985-8c31-4d87-b3cb-ea6c5220a66d)

## How To Test

1. Start anvil:
```bash
make anvil_start_with_block_time
```
2. Start the telemetry server in interactive mode:
```bash
make telemetry_run_db
make telemetry_ecto_migrate
cd telemetry_api
ALIGNED_CONFIG_FILE="../contracts/script/output/devnet/alignedlayer_deployment_output.json" OPERATOR_FETCHER_WAIT_TIME_MS=5000 ENVIRONMENT=devnet RPC_URL=http://localhost:8545 iex -S mix phx.server
```
3. Run metrics:
```bash
make run_metrics
```
4. Initialize different operator names from the telemetry terminal:
```Elixir
TelemetryApi.PrometheusMetrics.initialize_operator_metrics("gaston")
```
4. Call the `missing_operator` method from the telemetry terminal with the initialized operator names:
```Elixir
TelemetryApi.PrometheusMetrics.missing_operator("gaston")
```
6. Go to the aggregator-batcher dashboard in `localhost:3000` and you should see the dashboard with the values.

### Test also the full flow:
1. Start anvil:
```bash
make anvil_start_with_block_time
```
2. Start batcher:
```bash
make batcher_start_local
```
3. Go to `config-files/config-aggregator.yaml` and reduce the `bls_service_task_timeout`.
4. Start the aggregator:
```bash
make aggregator_start
```
7. Register and start the operator:
```bash
make operator_register_and_start
```
8. Send tasks:
```bash
make batcher_send_burst_groth16
```

9. Kill the operator and watch the panel.

## Type of change


## Checklist

  - [Link to the open issue addressing it]()
